Tests/Unit/NetworkServer/ConcentratorDeduplicationTest.cs

```diff
@@ -91,29 +91,32 @@ public void When_Data_Message_Encountered_Should_Find_Duplicates_For_Different_D
             // act/assert
             Assert.Equal(expectedResult, this.concentratorDeduplication.CheckDuplicateData(this.dataRequest, this.loRaDevice));
             Assert.Equal(1, this.cache.Count);
-            var key = ConcentratorDeduplication.CreateCacheKey(this.dataPayload);
+            var key = ConcentratorDeduplication.CreateCacheKey(this.dataPayload, this.loRaDevice);
             Assert.True(this.cache.TryGetValue(key, out var foundStation));
             Assert.Equal(station1Eui, foundStation);
         }
 
+        public static IEnumerable<object[]> CreateKeyDataMessagesTheoryData =>
```

Comment:
I think if you're going to go with theory data this way, you might as well make it strong-typed all the way:

```c#
internal static TheoryData<DevEui, Mic, ushort, string?> CreateKeyDataMessagesTheoryData =>
    TheoryDataFactory.From((new DevEui(0), new Mic(0), (ushort)0, null),
                           (new DevEui(0), new Mic(0), (ushort)0, "1"), // a non-relevant field should not influence the key
                           (new DevEui(0x1010101010101010UL), new Mic(0), (ushort)0, null),
                           (new DevEui(0), new Mic(1), (ushort)0, null),
                           (new DevEui(0), new Mic(0), (ushort)1, null));

[Theory]
[MemberData(nameof(CreateKeyDataMessagesTheoryData))]
internal void CreateKeyMethod_Should_Return_Expected_Keys_For_Different_Data_Messages(DevEui devEui, Mic mic, ushort frameCounter, string? fieldNotUsedInKey = null)
{
    var options = fieldNotUsedInKey ?? string.Empty;
    using var testDevice = new LoRaDevice(this.simulatedABPDevice.DevAddr, devEui.ToString(), this.connectionManager);

    var payload = new LoRaPayloadDataLns(this.dataPayload.DevAddr, new MacHeader(MacMessageType.ConfirmedDataUp),
                                         frameCounter, options, "payload", mic);

    var expectedKey = new ConcentratorDeduplication.DataMessageKey(devEui, mic, frameCounter);
    Assert.Equal(expectedKey, ConcentratorDeduplication.CreateCacheKey(payload, testDevice));
}
```

However, it seems odd to me that by exposing the key records (which I didn't in my original suggestion), you see that the last line is essentially _testing the implementation_ by re-creating it in the test code.

Reply:
There is 1 issue and 1 style-preference I would do differently from the committed sample: 
- Issue: MemberData need to be public otherwise we get a runtime exception when running the test
- I would rather we have all the test data as parameterized data and not only the inputs. 

Instead and in order to go more type-safe, see 656944045ea91c722afb20031f7e1915691a18fc
	- I added an overload in the `TheoryDataFactory` that takes 5 parameters 
	- When the `MemberData` is public, then the record type also needs to be public as well (otherwise CS0053 inconsistent accessibility) This would make the type globally visible which I think we should not do only for a test. Instead I am using object just for this parameter and still have type safety for the rest of the parameters + an implicit conversion at runtime for the record type. I think this is the best way in order not to make the record public and provide more type-safety.